### PR TITLE
Correctly caclculate refseq mismatch offset for seq edits between CDS…

### PR DESCRIPTION
**Pull request by aparton**
… and CDNA starts

The variant NC_000012.11:g.103249104C>A was annotated as HGVSc: NM_000277.1:*c.517*G>T in 99. The new RefSeq mismatch code overcorrected and gave c.515G>T. This was due to a seq edit in between the cdna_start and cds_start being considered when it shouldn't have been.

The expected HGVS output is  NM_000277.3:*c.516*G>T 

Edit: Additionally, this will also correctly shift XM_ transcripts with associated seq_edits

### CODE REVIEW

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm [hunk 1]
```diff
@@ -1417,15 +1417,15 @@ sub hgvs_transcript {
   my @edit_attrs = grep {$_->code =~ /^_rna_edit/} @attribs;
 
   my $misalignment_offset = 0;
-  $misalignment_offset = $self->get_misalignment_offset(\@edit_attrs) if (scalar(@edit_attrs) && substr($tr->stable_id, 0,3) eq 'NM_');
+  $misalignment_offset = $self->get_misalignment_offset(\@edit_attrs) if (scalar(@edit_attrs) && (substr($tr->stable_id, 0,3) eq 'NM_' || substr($tr->stable_id, 0,3) eq 'XM_'));
```
Comment (ima23) [inline modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: The example is NM, is the addition of `XM_` another bug that is fixed in this PR? If so, could this please be added in the PR comment. Thanks.

File: modules/t/transcriptVariationAdaptor.t [hunk 2]
```diff
@@ -350,4 +352,46 @@ my $msc_2_expected = 'test_consequence_a';
...
+my @tvas = @{ $tv->get_all_alternate_TranscriptVariationAlleles };
+$tvas[0]->{pre_consequence_predicates}->{exon} = 1;
+
+$tv->cds_start(1234);
+$tv->cds_end(1233);
+$tv->cdna_start(1000);
+$tv->transcript->{cdna_coding_start} = 234;
+
+## Coordinate within HGVS matches cds_start as these values take mismatch into account 
+ok($tv->hgvs_transcript->{A} eq 'NM_001270408.1:c.1234N>A', 'Refseq HGVS mismatch calculated');
```
Comment (ima23) [inline modules/t/transcriptVariationAdaptor.t]: Why is the reported hgvs containing 'N'? Could a comment be added or the PR example used here instead: `NC_000012.11:g.103249104C>A` leading to correct `NM_000277.3:c.516G>T`?

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm [hunk 3]
```diff
@@ -1417,15 +1417,15 @@ sub hgvs_transcript {
...
   
   if ($vf->var_class eq 'SNP' && defined($self->{pre_consequence_predicates}) && $self->{pre_consequence_predicates}->{exon} && defined($tv->cds_start) && defined($tv->cds_end)) {
```
Comment (ima23) [inline modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: Currently the test sets `{pre_conseuqence_predicates}->{exon}` which triggers this if. Where is this value expected to be set if a pipeline is run?

**Comment by ima23** [comment]
@aparton: the PR has a merge conflict, can you please have a look? Thanks.

File: modules/t/transcriptVariationAdaptor.t (hunk 2 above)
Comment (aparton) [reply modules/t/transcriptVariationAdaptor.t]: The test database does not include sequence data for the region the test is looking in. The N isn't relevant to what we're testing for, so I just left it.

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm (hunk 1 above)
Comment (aparton) [reply modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: Thanks, comment added

File: modules/t/transcriptVariationAdaptor.t [hunk 4]
```diff
@@ -350,4 +352,46 @@ my $msc_2_expected = 'test_consequence_a';
...
+$tv = Bio::EnsEMBL::Variation::TranscriptVariation->new(
+    -transcript     => $tr,
+    -variation_feature  => $vf_refseq,
+    -adaptor      => $trv_ad,
+    -no_shift     => 1,
+);
+
+## Setting CDS coordinates allows us to set up niche case where mismatches are calculated
+my @tvas = @{ $tv->get_all_alternate_TranscriptVariationAlleles };
+$tvas[0]->{pre_consequence_predicates}->{exon} = 1;
```
[the remark this answers is not in the document]
Comment (aparton) [reply modules/t/transcriptVariationAdaptor.t]: Thanks, comment added

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm (hunk 3 above)
Comment (aparton) [reply modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: Thanks, I've added a comment

File: modules/t/transcriptVariationAdaptor.t [hunk 5]
```diff
@@ -33,6 +33,7 @@ my $multi = Bio::EnsEMBL::Test::MultiTestDB->new();
 
 my $vdb = $multi->get_DBAdaptor('variation');
 my $db  = $multi->get_DBAdaptor('core');
+my $ofdb  = $multi->get_DBAdaptor('otherfeatures');
```
[the remark this answers is not in the document]
Comment (aparton) [reply modules/t/transcriptVariationAdaptor.t]: In an ideal world, yes, however I wont be able to get that change in before branching.

File: modules/t/transcriptVariationAdaptor.t [hunk 6]
```diff
@@ -350,4 +352,46 @@ my $msc_2_expected = 'test_consequence_a';
...
+$tv->cdna_start(1000);
+$tv->transcript->{cdna_coding_start} = 234;
+
+## Coordinate within HGVS matches cds_start as these values take mismatch into account 
+ok($tv->hgvs_transcript->{A} eq 'NM_001270408.1:c.1234N>A', 'Refseq HGVS mismatch calculated');
+
+## Misalignment offset calculated from transcript edits recognises insertion of 4BP
+my @attribs = @{$tr->get_all_Attributes()};
+my @edit_attrs = grep {$_->code =~ /^_rna_edit/} @attribs;
+ok($tvas[0]->get_misalignment_offset(\@edit_attrs) == 4, 'Misalignment offset');
```
[the remark this answers is not in the document]
Comment (aparton) [reply modules/t/transcriptVariationAdaptor.t]: Yeah, that's fine. This is just an additional test, the previous hgvs would be different without this change.

File: modules/t/transcriptVariationAdaptor.t (hunk 4 above)
[the remark this answers is not in the document]
Comment (ima23) [reply modules/t/transcriptVariationAdaptor.t]: Is the second line start ('then') needed?

File: modules/t/transcriptVariationAdaptor.t (hunk 5 above)
[the remark this answers is not in the document]
Comment (ima23) [reply modules/t/transcriptVariationAdaptor.t]: Sure. Could you add a jira ticket please so it's update with next occasion?

**Comment by ima23** [comment]
merged to `master` and `release/100`

**Pull request merged**